### ISSUE TITLE
performance drop while animating height using `interpolate` but works fine with `interpolateNode`

### ISSUE THREAD

**Issue by pSapien** [software-mansion/react-native-reanimated#2106]
## Description

There is a considerable frame drop while animating height with `interpolate` API. But seems to work smoothly with `interpolateNode`.

### Questions 

1. What is the difference between `interpolate` and `interpolateNode`?
2. Is there any case where one should be used over the other? Any tradeoffs to consider.
3. Should I better off use `interpolateNode` here instead of `interpolate`?

```
import * as React from 'react';
import { Text, View, StyleSheet, Dimensions, Image } from 'react-native';
import Constants from 'expo-constants';
import Animated, {
  useSharedValue,
  useAnimatedScrollHandler,
  Extrapolate,
  useAnimatedStyle,
  interpolate,
  interpolateNode,
} from "react-native-reanimated";

export const HEADER_IMAGE_HEIGHT = Dimensions.get("window").width / 3

const styles = StyleSheet.create({
  image: {
    position: "absolute",
    top: 0,
    left: 0,
    width: '100%',
    resizeMode: "cover",
  },
});

const IMAGE_URI = 'https://i.pinimg.com/originals/a4/1a/e5/a41ae5ff09234422737d3899cc895184.jpg'

const touchX = 100;

const heightInputRange = [-touchX, 0];
const heightOutputRange = [HEADER_IMAGE_HEIGHT + touchX, HEADER_IMAGE_HEIGHT];
const heightAnimConfig = {
  extrapolateRight: Extrapolate.CLAMP,
}

const topInputRange = [0, touchX];
const topOutputRange = [0, -touchX];
const topAnimConfig = {
  extrapolateLeft: Extrapolate.CLAMP,
}

export default function App() {
  // const scrollY = useSharedValue(0)

  // const scrollHandler = useAnimatedScrollHandler({
  //   onScroll: (e) => {
  //     scrollY.value = e.contentOffset.y;
  //   },
  // });

 // const dragStyles = useAnimatedStyle(() => {
  //   return {
  //     height: interpolate(scrollY.value, heightInputRange, heightOutputRange, heightAnimConfig),
  //     top: interpolate(scrollY.value, topInputRange, topOutputRange, topAnimConfig),
  //   }
  // })

  const scrollY = new Animated.Value(0);
  const scrollHandler = Animated.event([
    {
      nativeEvent: {
        contentOffset: {
          y: scrollY
        }
      }
    }
  ])
  const animStyles = {
    height: interpolateNode(scrollY, { 
      inputRange: heightInputRange,
      outputRange: heightOutputRange,
      extrapolateRight: Extrapolate.CLAMP,
    }),
    top: interpolateNode(scrollY, { 
      inputRange: topInputRange,
      outputRange: topOutputRange,
      extrapolateLeft: Extrapolate.CLAMP,
    }),
  }

  return (
    <View style={{ flex: 1 }}>
      <Animated.Image
        style={[styles.image, animStyles]}
        source={{ uri: IMAGE_URI }}
      />

      <Animated.ScrollView
        scrollEventThrottle={1}
        style={StyleSheet.absoluteFill}
        onScroll={scrollHandler}
      >
        <View style={{ height: HEADER_IMAGE_HEIGHT + 200 }} />
        {Array.from({ length: 50 }).map((v, idx) => {
          return (
            <View key={idx}>
              <Text>{idx}</Text>
            </View>
          )
        })}
      </Animated.ScrollView>
    </View>
  );
}
```

#### Expected Behaviour:- 
It should work smoothly with the new `interpolate` API, shouldn't it?

#### Actual Behaviour:-
The performance drops while using with the new `interpolate` API?

#### Step to Reproduce

1. Open the snack link down below.
2. Try dragging the header image down to see the performance drop using the new API.

#### Snack link 
https://snack.expo.io/@sapien/74bdc2

Package versions :-
------------------------

-   react: "0.64.1"
-   react-native: "17.0.1"
-   react-native-reanimated: "^2.2.0"

Devices used :- iPhone 8 and iPhone 11

Please Note:- The snack here uses v2.1.0 as its expo doesn't allow me to update the version. In my project, I'm using v2.2.0 though.

**Comment by andreialecu** [comment]
I'd also like to know this, the migration docs mention that in v2 we should use `interpolateNode`:

> 1. interpolate renamed to interpolateNode
When using interpolate imported directly from react-native-reanimated v1, in v2 you should use interpolateNode instead. If you were using a class member method AnimatedValue.interpolate, no change is necessary.

https://github.com/software-mansion/react-native-reanimated/blob/master/docs/versioned_docs/version-2.3.0-alpha.1/migration.md

However, `interpolateNode` is not documented at all. It also seems to return `Animated.Node<number>` instead of `number`, which makes subsequent calculations impossible when part of worklets. 

`interpolate` seems to be documented as part of reanimated v1:
https://docs.swmansion.com/react-native-reanimated/docs/1.x.x/nodes/interpolate/

But the function signature seems to match the one from `interpolateNode` instead.

**Comment by mrousavy** [comment]
Yes, as of v2 you should use `interpolateNode` to interpolate animated _nodes_ (the v1 way of animations with reanimated), and `interpolate` to interpolate _Shared Values_ (the v2 way of animations with reanimated aka in worklets).

**Comment by jmysliv** [comment]
Hi @pSapien, I've tested it with reanimated `2.3.0-alpha.2` and I didn't encounter any frame drop. If by frame drop you meant a rapid change of the height, it happened because you used `extrapolateRight: Extrapolate.CLAMP`, so height is set to `HEADER_IMAGE_HEIGHT` till you scroll enough. If you remove it from config it should work fine.

> However, `interpolateNode` is not documented at all. It also seems to return `Animated.Node<number>` instead of `number`, which makes subsequent calculations impossible when part of worklets.
> 
> `interpolate` seems to be documented as part of reanimated v1:
> https://docs.swmansion.com/react-native-reanimated/docs/1.x.x/nodes/interpolate/

Tomorrow I'm going to update docs to make it more clear.

**Comment by jmysliv** [comment]
Hi, after a better investigation, I found out that `extrapolationRight` was applied in the `interpolate` function inside the range and not outside. That is why after removing `extrapolationRight`, it worked fine. I fixed it and updated the docs here #2321 .

**Comment by jmysliv** [comment]
Fixed by #2356
